doc/dd_post.1.rst

=========
 DD_Post
=========

------------------------------------------------
Publish the Availability of a File to Subcribers
------------------------------------------------

:Manual section: 1 
:Date: Aug 2015
:Version: 0.0.1
:Manual group: MetPx Sarracenia Suite


SYNOPSIS
========

**dd_post** [ *-u|--url url* ] [ *-b|--broker broker* ]...[ *OPTIONS* ]

DESCRIPTION
===========

**dd_post** posts the availability of a file by creating an announcment.
Subscribers use `dd_subscribe <dd_subscribe.1.html>`_ to consume the announcement and 
download the file.  To make files available to subscribers, **dd_post** sends 
the announcements to an AMQP server, also called a broker.  Format of argument 
to the *broker* option:: 

       [amqp|amqps]://[user[:password]@]host[:port][/vhost]

The [*-u|--url url*] option specifies the location 
from which subscribers will download the file.  There is usually one post per file.
Format of argument to the *url* option::

       [ftp|http|sftp]://[user[:password]@]host[:port]//absolute_path_to_the/filename
       or
       [ftp|http|sftp]://[user[:password]@]host[:port]/relative_path_to_the/filename
       or
       file://absolute_path_to_the/filename

The double-slash at the beginning of the path marks it as absolute, whereas a single
slash is relative to a *document_root* provided as another option.
An example invocation of *dd_post*::

 dd_post -u sftp://stanley@mysftpserver.com//data/shared/products/foo -b amqp://broker.com

By default, dd_post reads the file /data/shared/products/foo and calculates its checksum.
It then builds a post message, logs into broker.com as user 'guest' (default credentials)
and sends the post  to defaults vhost '/' and default exchange 'amq.topic'.
A subscriber can download the file /data/shared/products/foo by authenticating as user stanley
on mysftpserver.com using the sftp protocol to broker.com assuming he has proper credentials.
The output of the command is as follows ::

 [INFO] v02.post.data.shared.products.foo  '20150813161959.854 sftp://stanley@mysftpserver.com/ /data/shared/products/foo'
              source=guest sum=d,82edc8eb735fd99598a1fe04541f558d parts=1,4574,1,0,0

In MetPX-Sarracenia, each post is published under a certain topic.
The log line starts with '[INFO]', followed by the **topic** of the
post. Topics in *AMQP* are fields separated by dot. The complete topic starts with
a topic_prefix (see option)  version *V02*, an action *post*,
followed by a subtopic (see option) here the default, the file path separated with dots
*data.shared.products.foo*

.. NOTE::
  FIXME: the topic does not contain the user?  Should it contain the source user? some docs say yes.

The second field in the log line is the message notice.  It consists of a time 
stamp *20150813161959.854*, and the source url of the file in the last 2 fields.

the rest of the information comes message headers, consisting of key-value pairs.
the first header should is *source=guest* indicating the user used to authenticate to the broker.
The *sum=d,82edc8eb735fd99598a1fe04541f558d* header gives file fingerprint (or checksum
) information.  Here, *d* means md5 checksum performed on the data, and *82edc8eb735fd99598a1fe04541f558d*
is the checksum value. The *parts=1,4574,1,0,0* state that the file is available in 1 part of 4574 bytes
(the filesize.)  The remaining *1,0,0* is not used for transfers of files with only one part.

Another example::

 dd_post -dr /data/web/public_data -u http://dd.weather.gc.ca/bulletins/alphanumeric/SACN32_CWAO_123456 -b amqp://broker.com

By default, dd_post reads the file /data/web/public_data/bulletins/alphanumeric/SACN32_CWAO_123456
(concatenating the document_root and relative path of the source url to obtain the local file path)
and calculates its checksum. It then builds a post message, logs into broker.com as user 'guest'
(default credentials) and sends the post to defaults vhost '/' and exchange 'amq.topic'

A subscriber can download the file http://dd.weather.gc.ca/bulletins/alphanumeric/SACN32_CWAO_123456 using http
without authentication on dd.weather.gc.ca.

.. NOTE::
  FIXME: the amq.topic ... is this correct? xpublic? validate.


ARGUMENTS AND OPTIONS
=====================

**[-b|--broker <broker>]**

  the broker to which the post is sent.

**[-c|--config <configfile>]**

  Any command line arguments has a corresponding long version starting with '--'.
  For example *-u* has the long form *--url*. You can also specify
  this option in a configuration file shall you need it. To do so, you simply
  use the long form without the '--', and put its value separated by a space.
  In a configuration file the right syntax to set the url is :

**url <url>** 

  The *config* option is no exception... and if used the content of this
  other specified file will have its options processed.

**[-dr|--document_root <path>]**

  The *document_root* option supplies the directory path that,
  when combined with the relative one from *url*, 
  gives the local absolute path to the data file to be posted.

**[-ex|--exchange <exchange>]**

  By default, the exchange used is amq.topic. This exchange is provided on broker
  for general usage. It can be overwritten with this *exchange* option

.. NOTE::
   FIXME: default is wrong?  figure out how often used?

**[-f|--flow <string>]**

  An arbitrary label that allows the user to identify a specific flow.
  The flow string is sets in the amqp message header.  By default, there is no flow.

**[-h|-help|--help**

  Display program options.

**[-rn|--rename <path>]**

  With the *rename*  option, the user can suggest a destination path to its files. If the given
  path ends with '/' it suggests a directory path...  If it doesn't, the option specifies a file renaming.

**[-tp|--topic_prefix <key>]**

  *Not usually used*
  By default, the topic is made of the default topic_prefix : version *V02*, an action *post*,
  followed by the default subtopic: the file path separated with dots (dot being the topic separator for amqp).
  You can overwrite the topic_prefix by setting this option.

**[-sub|--subtopic <key>]**

The subtopic default can be overwritten with the *subtopic* option.

**[-u|--url <url>]**

*url* is the actual download url to be
used by the subscribers.

ADVANCED OPTIONS
================

**[-p|--parts <value>]**

The user can suggest how to download a file.  By default it suggests to download the entire file.
In this case, the amqp message header will have an entry parts with value '1,filesize_in_bytes'.
To suggest to download a file in blocksize of 10Mb, the user can specify *-p i,10M*. *i* stands for
"inplace" and means to put the part directly into the file.  *-p p,10M* suggests the same blocksize but to put the part
in a separate filepart. If the *blocksize* is bigger than the filesize, the program will fall back to the default.
There will be one post per suggested part.

The value of the *blocksize*  is an integer that may be followed by  letter designator *[B|K|M|G|T]* meaning:
for Bytes, Kilobytes, Megabytes, Gigabytes, Terabytes respectively.  All theses references are powers of 2.

When suggesting parts, the value put in the amqp message header varies.
For example if headers[parts] as value 'p,256,12,11,4' it stands for :
*p* suggesting part, a blocksize in bytes *256*,
the number of block of that size *12*, the remaining bytes *11*, 
and the current block *4*,

**[-sum|--sum <string>]**

All file posts include a checksum.  It is placed in the amqp message header will have as an
entry *sum* with default value 'd,md5_checksum_on_data'.
The *sum* option tell the program how to calculate the checksum.
It is a comma separated string.  Valid checksum flags are ::

    [0|n|d|c=<scriptname>]
    where 0 : no checksum... value in post is 0
          n : do checksum on filename
          d : do md5sum on file content

Then using a checksum script, it must be registered with the switch, so that consumers
of the postings have access to the algorithm.


DEVELOPER SPECIFIC OPTIONS
==========================

**[-debug|--debug]**

Active if *-debug|--debug* appears in the command line... or
*debug* is set to True in the configuration file used.

**[-r|--randomize]**

Active if *-r|--randomize* appears in the command line... or
*randomize* is set to True in the configuration file used.
If there are several posts because the file is posted
by block because the *blocksize* option was set, the block 
posts are randomized meaning that the will not be posted
ordered by block number.

**[-rr|--reconnect]**

Active if *-rc|--reconnect* appears in the command line... or
*reconnect* is set to True in the configuration file used.
*If there are several posts because the file is posted
by block because the *blocksize* option was set, there is a
reconnection to the broker everytime a post is to be sent.

SEE ALSO
========

`dd_log(7) <dd_log.7.html>`_ - the format of log messages.

`dd_post(7) <dd_post.7.html>`_ - the format of announcement messages.

`dd_sara(1) <dd_sara.1.html>`_ - Subscribe, Acquire, and ReAdvertise tool.

`dd_subscribe(1) <dd_subscribe.1.html>`_ - the http-only download client.

`dd_watch(1) <dd_watch.1.html>`_ - the directory watching daemon.





sara/dd_post.py
#!/usr/bin/python3
#
# This file is part of sarracenia.
# The sarracenia suite is Free and is proudly provided by the Government of Canada
# Copyright (C) Her Majesty The Queen in Right of Canada, Environment Canada, 2008-2015
#
# Questions or bugs report: dps-client@ec.gc.ca
# sarracenia repository: git://git.code.sf.net/p/metpx/git
# Documentation: http://metpx.sourceforge.net/#SaraDocumentation
#
# dd_post.py : python3 program allowing users to post an available product
#
#
# Code contributed by:
#  Michel Grenier - Shared Services Canada
#  Jun Hu         - Shared Services Canada
#  Last Changed   : Sep 22 10:41:32 EDT 2015
#  Last Revision  : Sep 22 10:41:32 EDT 2015
#
########################################################################
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful, 
#  but WITHOUT ANY WARRANTY; without even the implied warranty of 
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
#
#

import os,random,sys

try :
         from dd_amqp         import *
         from dd_config       import *
         from dd_message      import *
         from dd_util         import *
except :
         from sara.dd_amqp    import *
         from sara.dd_config  import *
         from sara.dd_message import *
         from sara.dd_util    import *

class dd_post(dd_config):

    def __init__(self,config=None,args=None):
        dd_config.__init__(self,config,args)
        self.configure()

    def check(self):

        if self.url == None :
           self.logger.error("url required")
           sys.exit(1)

        self.chkclass = Checksum()
        self.chkclass.from_list(self.sumflg)
        self.chksum = self.chkclass.checksum

        # sara exchange default value is xs_username
        # username being the broker's

        if self.exchange == None :
           self.exchange = 'xs_%s' % self.broker.username

        self.msg = dd_message(self.logger)
        self.msg.set_exchange(self.exchange)
        self.msg.set_flow(self.flow)
        self.msg.set_flow(self.flow)
        self.msg.set_to_clusters(self.to_clusters)

    def close(self):
        self.hc_post.close()

    def configure(self):

        # installation general configurations and settings

        self.general()

        # defaults general and proper to dd_post

        self.defaults()

        # arguments from command line

        self.args(self.user_args)

        # config from file

        self.config(self.user_config)

        # verify all settings

        self.check()


    def connect(self):

        self.hc_post      = HostConnect( logger = self.logger )
        self.hc_post.set_url( self.broker )

        # dd_post : no loop to reconnect to broker

        if self.program_name == 'dd_post' :
           self.hc_post.loop = False
                                   
        self.hc_post.connect()

        self.pub    = Publisher(self.hc_post)
        self.pub.build()

    def help(self):
        self.logger.info("Usage: %s -u <url> -b <broker> ... [OPTIONS]\n" % self.program_name )
        self.logger.info("OPTIONS:")
        self.logger.info("-b   <broker>          default:amqp://guest:guest@localhost/")
        self.logger.info("-c   <config_file>")
        self.logger.info("-dr  <document_root>   default:None")
        if self.program_name == 'dd_watch' : self.logger.info("-e   <events>          default:IN_CLOSE_WRITE\n")
        self.logger.info("-ex  <exchange>        default:xs_\"broker.username\"")
        self.logger.info("-f   <flow>            default:None\n")
        self.logger.info("-h|--help\n")
        self.logger.info("-l   <logpath>         default:stdout")
        self.logger.info("-p   <parts>           default:1")
        self.logger.info("-tc  <to_clusters>     default:None")
        self.logger.info("-tp  <topic_prefix>    default:v02.post")
        self.logger.info("-sub <subtopic>        default:'path.of.file'")
        self.logger.info("-rn  <rename>          default:None")
        self.logger.info("-sum <sum>             default:d")
        self.logger.info("DEBUG:")
        self.logger.info("-debug")
        self.logger.info("-r  : randomize chunk posting")
        self.logger.info("-rr : reconnect between chunks")

    def instantiate(self,i=0):
        self.instance = i
        self.setlog()

    def posting(self):

        filepath = '/' + self.url.path.strip('/')

        # urllib keeps useless repetitive '/' so rebuild url smartly
        if filepath != self.url.path :
           if self.document_root == None and self.url.scheme[-3:] == 'ftp' :
              filepath = '/' + filepath
           urlstr   = self.url.scheme + '://' + self.url.netloc + filepath
           self.url = urllib.parse.urlparse(urlstr)

        # check abspath for filename

        filepath = self.url.path
        if self.document_root != None :
           if str.find(filepath,self.document_root) != 0 :
              filepath = self.document_root + os.sep + filepath
              filepath = filepath.replace('//','/')

        # verify that file exists

        if not os.path.isfile(filepath) and self.event != 'IN_DELETE' :
           self.logger.error("File not found %s " % filepath )
           return False

        # verify part file... if it is ok

        if self.partflg == 'p' :
           ok,message = self.verify_p_file(filepath)
           if not ok:
              self.logger.error("partflg set to p but %s for file  %s " % (message,filepath))
              return False

        # rename path given with no filename

        rename = self.rename
        if self.rename != None and self.rename[-1] == os.sep :
           rename += os.path.basename(self.url.path)

        # make sure a part file has a part name
        if self.rename != None and self.partflg == 'p' :
           if not self.msg.suffix in rename :
              rename += self.msg.suffix

        self.msg.set_rename(rename)

        #

        self.logger.info("broker(%s) user(%s) vhost(%s)  exchange(%s)" % \
        (self.broker.hostname,self.broker.username,self.broker.path,self.exchange) )

        # ==============
        # delete event...
        # ==============

        if self.event == 'IN_DELETE' :
           self.msg.set_parts(None)
           self.msg.set_sum(sumflg='R')
           self.publish()
           return

        # ==============
        # p partflg special case
        # ==============

        if self.partflg == 'p' :
           blocksize, block_count, remainder, current_block, sum_data = self.p_chunk
           self.msg.set_parts('p', blocksize, block_count, remainder, current_block)
           self.msg.set_sum(self.sumflg,sum_data)
           self.publish()
           return

        # ==============
        # Chunk set up
        # ==============

        chunk  = Chunk(self.blocksize,self.chksum,filepath)
        N      = chunk.get_Nblock()

        # ==============
        # Randomize
        # ==============

        rparts = list(range(0,N))

        # randomize chunks
        if self.randomize and N>1 :
           i = 0
           while i < N/2+1 :
               j         = random.randint(0,N-1)
               tmp       = rparts[j]
               rparts[j] = rparts[i]
               rparts[i] = tmp
               i = i + 1

        # ==============
        # loop on chunk
        # ==============

        i  = 0
        while i < N:

            # build message
 
            c = chunk.get( rparts[i] )
            blocksize, block_count, remainder, current_block, sum_data = c

            self.msg.set_parts(self.partflg, blocksize, block_count, remainder, current_block)
            self.msg.set_sum(self.sumflg,sum_data)

            self.publish()

            i = i + 1

            # reconnect ?
            if self.reconnect and i<N :
               self.logger.info("Reconnect")
               self.hc_post.reconnect()

    def publish(self):
        if self.subtopic == None :
           self.msg.set_topic_url(self.topic_prefix,self.url)
        else :
           self.msg.set_topic_usr(self.topic_prefix,self.subtopic)

        self.msg.set_notice(self.url)
        self.msg.set_hdrstr()
        self.logger.info("%s '%s' %s" % (self.msg.topic,self.msg.notice,self.msg.hdrstr))
        ok = self.pub.publish( self.msg.exchange, self.msg.topic, self.msg.notice, self.msg.headers )
        if not ok : sys.exit(1)
        self.logger.info("published")

    def verify_p_file(self,filepath):
        ok,message = self.msg.verify_part_suffix(filepath)

        if ok :
           self.p_chunk = (self.msg.chunksize, self.msg.block_count, self.msg.remainder, self.msg.current_block, self.msg.checksum)

        return ok,message

    def watching(self, fpath, event ):

        self.event = event

        if self.document_root != None :
           fpath = fpath.replace(self.document_root,'')
           if fpath[0] == '/' : fpath = fpath[1:]

        url = self.url
        self.url = urllib.parse.urlparse('%s://%s/%s'%(url.scheme,url.netloc,fpath))
        self.posting()
        self.url = url

    def watchpath(self ):

       watch_path = self.url.path
       if watch_path == None : watch_path = ''

       if self.document_root != None :
          if not self.document_root in watch_path :
             watch_path = self.document_root + os.sep + watch_path

       watch_path = watch_path.replace('//','/')

       if not os.path.exists(watch_path):
          self.logger.error("Not found %s " % watch_path )
          sys.exit(1)

       if os.path.isfile(watch_path):
          self.logger.info("Watching file %s " % watch_path )

       if os.path.isdir(watch_path):
          self.logger.info("Watching directory %s " % watch_path )
          if self.rename != None and self.rename[-1] != '/' and 'IN_CLOSE_WRITE' in self.events:
             self.logger.warning("renaming all modified files to %s " % self.rename )

       return watch_path


# ===================================
# MAIN
# ===================================

def main():

    post = dd_post(config=None,args=sys.argv[1:])

    try :
             post.instantiate()
             post.connect()
             post.posting()
             post.close()
    except :
             (stype, value, tb) = sys.exc_info()
             post.logger.error("Type: %s, Value:%s\n" % (stype, value))
             sys.exit(1)


    sys.exit(0)

# =========================================
# direct invocation
# =========================================

if __name__=="__main__":
   main()

